File: antia_numerical_methods@Beliavsky/c/xmp/binom.c
/*	To simulate a binomial distribution */

#include <stdio.h>
#include <math.h>

double gamma(double xg);
double gammaln(double xg);
double gammap(double a,double x);
int iranbin(double *seed, int n, double p, double c[]);

main()
{
	int i,n,ii,nu,num,ip[200];
	double p,p0,r,rn,seed,chi,prob,c[200];

	rn=1e4; num=rn;
	printf(" Type n = No. of trials (<200)\n");
	printf(" and p = probability of event in binomial distribution\n");
	scanf("%d %le",&n,&p);
	if(n>=200) return;
	seed=-1;
	for(i=0; i<200; ++i) ip[i]=0.0;
/*	Generate num random numbers with binomial distribution and
        place them in the appropriate bin */
	for(i=0; i<num; ++i){
		ii=iranbin(&seed,n,p,c);
		ip[ii]=ip[ii]+1;
	}

/*	Apply the Chi-square test to check the distribution
        p0 is the expected probability
        This will not work if initial P0 gives underflow	*/
	p0=pow(1.0-p,n); chi=0.0;
	printf("       I        Prob(I)\n");
	for(i=0; i<=n; ++i) {
		printf(" %d %e %e \n",i,ip[i]/rn,p0);
		r=p0*rn-ip[i]; chi=chi+r*r/(p0*rn);
		p0=p0*(n-i)*p/((i+1)*(1.0-p));
	}
	nu=n;
	printf(" Chi-square = %e    Degree of freedom = %d \n",chi,nu);
/*	If NUM is increased and approximation to Gaussian distribution
        is used in IRANBIN the test may fail if P is close to 1
        as the Gaussian distribution may not be a good approximation in
        that case. RMAX in IRANBIN may be increased to avoid use of
        Gaussian approximation */
	prob=1-gammap(nu/2.0,chi/2);
	printf(" Probability of finding a value of CHI or larger = %e\n",prob);
}

/*	To calculate Gamma function for any real value of XG
	Use GAMMALN for calculating the logarithm of Gamma function
	which may be useful for large arguments or when argument is
	close to a negative integer.

	Required functions : None
*/

#include <math.h>

/*	PIS is SQRT(2*PI) */
#define PI 3.14159265358979323846
#define PIS 2.5066282746310005024

double gamma(double xg)

{
	double x,gx,f,y,rmk,f1,x1,fn,fd;
	
/*	The coefficients for rational function approximations */
	double a[2] ={1.767971449569122937e+00,  2.909421117928672645e-01};
	double b[3] ={8.333333333333231537e-02,  1.445531763554246280e-01,
                      2.012779361583001035e-02};
	double a1[6] ={3.905731686764559737e+03,  2.204952264401381785e+03,
                      -1.932467485468849660e+03,  4.643360871045442213e+02,
                      -4.818088806916028754e+01,  1.896853765546068169e+00};
	double b1[7] ={3.918055655523400310e+03, -1.088116266563809683e+02,
                       8.203258626193993149e+02, -9.289402000761705906e+01,
                       6.521113026294866877e+01, -6.090618615608719044e+00,
                       1.475909104740280784e+00};
 
	x=fabs(xg);
	if(x>1000.0) {
/*	Use asymptotic approximation (Stirling formula) */
        gx=(1+1.0/(12*x)+1./(288*x*x)-139.0/(51840*x*x*x)-571.0/(2488320e0*x*x*x*x));
		f=pow(x,x-0.5)*exp(-x)*PIS*gx;
	}

	else if(x>8.0) {
/*	Use rational function approximation for Log(Gamma)  */
		y=1.0/(x*x);
        rmk=((b[2]*y+b[1])*y+b[0])/((a[1]*y+a[0])*y+1);
		f=pow(x,x-0.5)*exp(-x)*PIS*exp(rmk/x);
	}
	else if(x>=2.0) {
/*	Use rational function approximation for (Gamma) over [2,3]
	after translating the range if necessary */
		f1=1.0; x1=x;
		while(x1>3.0) {
			f1=f1*(x1-1.0);
			x1=x1-1.0;
		}

		if(x1==3) f=f1*2.0;
		else if(x1==2) f=f1;

		fn=(((((b1[6]*x1+b1[5])*x1+b1[4])*x1+b1[3])*x1+b1[2])*x1+b1[1])*x1+b1[0];
		fd=(((((a1[5]*x1+a1[4])*x1+a1[3])*x1+a1[2])*x1+a1[1])*x1+a1[0])*x1+1;
		f=f1*fn/fd;
	}
	else if(x>0.0) {
/*	Use rational function approximation for (Gamma) over [2,3]
	after translating the range if necessary */
		f1=1./x;
		x1=x+1.0;
		if(x<1) {f1=f1/x1; x1=x1+1.0;}
		if(x1==2) f=f1;

		fn=(((((b1[6]*x1+b1[5])*x1+b1[4])*x1+b1[3])*x1+b1[2])*x1+b1[1])*x1+b1[0];
		fd=(((((a1[5]*x1+a1[4])*x1+a1[3])*x1+a1[2])*x1+a1[1])*x1+a1[0])*x1+1;
		f=f1*fn/fd;
	}

	if(xg>0.0) return f;
	f=PI/(xg*sin(PI*x)*f);
	return f;
}
 

/*	To calculate the Logarithm of Gamma function for a real argument
 	For negative values it give ln(abs(Gamma(x)))
 
 	Required routines : None
*/

#include <math.h>

#define PI 3.14159265358979323846

double gammaln(double xg)

{
	double x,gx,f,y,rmk,f1,x1,fn,fd;
/*	PIL is LOG(SQRT(2*PI)) */
	double PI2L=0.918938533204672741;
	
/*	The coefficients for rational function approximations */
	double a[4] = {1.013142782275024216e-2,  7.645657825398191944e-1,
             3.381172379819461227e-4,  1.595363637547538209e-2};
	double b[4] ={0.0, 8.333333333338911768e-2, 8.442856404442060242e-4,
                 6.093603832366013704e-2};
	double a1[6] ={4.681163846241230144,  3.208225429683256526,
              5.145525793448859216e-1, 1.581117883959157936e-2,
             -6.398416804905407512e-5, 5.264566254181773919e-7};
	double b1[6] ={2.938038561191284576,  1.489364948862436743,
             -5.466291543917642961,  1.972497734170110410e-1,
              7.830146473241555157e-1, 5.756753067834747499e-2};
 
	x=fabs(xg);
	if(x>10.0) {
/*	Use asymptotic approximation (Stirling formula) */
		y=1.0/x;
		fn=((b[3]*y+b[2])*y+b[1])*y+b[0];
		fd=(((a[3]*y+a[2])*y+a[1])*y+a[0])*y+1;
		f=(x-0.5)*log(x)-x+PI2L+fn/fd;
	}

	else {
/*	Use rational function approximation for (Gamma) over [4,5]
	after translating the range if necessary */
		f1=1.0; x1=x;
		while(x1>5.0) {
			f1=f1*(x1-1.0);
			x1=x1-1.0;
		}
		while(x1<4.0) {
			f1=f1/(x1);
			x1=x1+1.0;
		}


		fn=((((b1[5]*x1+b1[4])*x1+b1[3])*x1+b1[2])*x1+b1[1])*x1+b1[0];
		fd=(((((a1[5]*x1+a1[4])*x1+a1[3])*x1+a1[2])*x1+a1[1])*x1+a1[0])*x1+1;
		f=log(f1)+fn/fd;
	}

	if(xg>0.0) return f;
	f=log(PI/(x*fabs(sin(PI*x))))-f;
	return f;
}
 

/*	To calculate the incomplete Gamma function P(a,x) for
	positive real arguments
	It returns a value of -1 for a.le.0 or x<0

	a : (input) Argument for the complete Gamma function
	x : (input) Upper limit of integration defining the incomplete
			Gamma function

	Required routines : gamma, gammaln
*/
 

#include <math.h>

double gamma(double xg);
double gammaln(double xg);

double gammap(double a,double x)
{
	double c[500],b[500],f,s1,s2,g1,t,c1,c2;
	int i;

	if(a<=0 || x<0) return -1.0; 

	if(x<3)
	{
/*	Use power series */
        	s1=1; t=a; i=0;
		while(i<500 && fabs(t/(a+i))>1.e-14)
		{
			i=i+1; t=-t*x/i;
			s1=s1+t/(a+i);
		}
		if(a<140) f=pow(x,a)*s1/gamma(a+1);
		else f=0.0;
	}
	else if(a<1.2*x)
	{
/*	Use continued fraction */
        	c[0]=1; b[0]=x; c[1]=1; b[1]=x+1-a;
		c2=c[1]/b[1]; c1=0.0; i=0;
		while(i<500 && fabs(c1-c2)>1.e-12)
		{
			i=i+2; c1=c2;
			c[i]=x*c[i-1]+(i/2)*c[i-2];
			b[i]=x*b[i-1]+(i/2)*b[i-2];
			c[i+1]=c[i]+(i/2+1-a)*c[i-1];
			b[i+1]=b[i]+(i/2+1-a)*b[i-1];
			if(fabs(b[i+1])>1.e200)
			{
				c[i]=c[i]/1.e200; c[i+1]=c[i+1]/1.e200;
				b[i]=b[i]/1.e200; b[i+1]=b[i+1]/1.e200;
			}
			if(fabs(b[i+1])<1.e-200)
			{
				c[i]=c[i]*1.e200; c[i+1]=c[i+1]*1.e200;
				b[i]=b[i]*1.e200; b[i+1]=b[i+1]*1.e200;
			}
			c2=c[i+1]/b[i+1];
		}
		g1=-x+a*log(x)+log(c2)-gammaln(a); f=1-exp(g1);
	}
	else
	{
/*	Use the power series for a>x */
		s2=1; t=1; i=0;
		while(i<500 && fabs(t)>1.e-14)
		{
			i=i+1; t=t*x/(a+i); s2=s2+t;
		}
		g1=-x+a*log(x)+log(s2)-gammaln(a+1); f=exp(g1);
	}
	return f;
}


/*	To generate random numbers with Binomial distribution 
        For large values of mean it approximates the distribution by a
        normal distribution with the same mean and standard deviation.
 
        SEED : (input/output) real seed, it should be negative during
               the first call and should not be modified between two
               calls with the same N, P. If a new value of N or P is used
               it should be reset to a negative value.
        N   :  (input) Number of trials in the binomial distribution
        P   :  (input) probability of the event
        C   :  (output) Real array of length N used to store the
               cumulative probability table for use in subsequent
               calculations. This array should not be modified by the
               user. This is used only if the mean N*P<RMAX and there
               is no underflow while calculating (1-P)**N
 
        Required routines : None

	If P is close to 1 the approximation by Gaussian may not be very
        good and in that case increasing RMAX may help although it will
        take longer time to generate the numbers.
*/


#include <math.h>

int iranbin(double *seed, int n, double p, double c[])
{
	int i;
	double am=2147483648.0,a=45875.0,ac=453816693.0,an=2147483647.0;
	double pi=3.14159265358979324,rmax=100;
	double rmu,sig,r1,rangau,p0,r;

	rmu=n*p; sig=sqrt(rmu*(1-p)); p0=pow(1-p,(double) n);

/*	For large mu or if P0 gives underflow use normal distribution */
	if(rmu > rmax || p0==0) {
/*	Initialise the seed during first call */
		if(*seed<0.0) *seed=fmod(fabs(*seed),am);
		r1=fmod(a*(*seed)+ac,am);
		if(*seed==0.0) *seed=0.1;
		rangau=sqrt(2.0*log(an/(*seed)))*cos(2.0*pi*r1/an);
		*seed=r1;
		i=rmu+sig*rangau+0.5;
		if(i<0) i=0; if(i>n) i=n;
		return i;
	}
	else {
/*       Initialise the array during the first call */
		if(*seed<0.0) {
			c[0]=p0;
			for(i=1; i<=n; ++i) {
				p0=p0*p*(n-i+1)/((1-p)*i);
				c[i]=c[i-1]+p0;
			}
			*seed=fabs(*seed);
		}

/*	generate the random number */
		*seed=fmod((*seed)*a+ac,am); r=(*seed)/an;
		for(i=0; i<=n; ++i) {
			if(r<c[i]) return i;
		}
		return n;
	}
}
	





	

	

File: projects/antia_numerical_methods@Beliavsky/f77/xmp16/binom.f
!       To simulate a binomial distribution
      PROGRAM RANBINOM
      IMPLICIT REAL*16(A-H,O-Z)
      DIMENSION C(200),IP(200)

51    FORMAT(I5,1P5D13.5)
52    FORMAT('   N =',i5,'    Probability =',F10.6,/,
     1        '    I    PROB(I) ')
53    FORMAT(/,' Chi-square = ',1PD13.5,'  Degree of freedom =',I4)
54    FORMAT(' Probability of finding a value of CHI or larger =',
     1               1PD13.5)

      RN=1Q4
      NUM=RN
      PRINT *,' TYPE N = No. of trials (< 200)'
      PRINT *,' and P = probability of event in binomial distribution'
      READ *,N,P
      IF(N.GE.200) STOP
      SEED=-1
      DO I=1,200
        IP(I)=0.0
      ENDDO
!       Generate NUM random numbers with binomial distribution and
!       place them in the appropriate bin
      DO I=1,NUM
        II=IRANBIN(SEED,N,P,C)
        IP(II+1)=IP(II+1)+1
      ENDDO

!       Apply the Chi-square test to check the distribution
!       p0 is the expected probability
!       This will not work if initial P0 gives underflow
      P0=(1.0-P)**N
      CHI=0.0
      WRITE(6,52) N,P
      NU=0
      DO I=1,N+1
        WRITE(6,51)I-1,IP(I)/RN,P0
        CHI=CHI+(P0*RN-IP(I))**2/(P0*RN)
        P0=P0*(N-I+1)*P/(I*(1.0-P))
        IF(IP(I).GT.0) NU=NU+1
      ENDDO
      NU=N
      WRITE(6,53) CHI,NU
!       If NUM is increased and approximation to Gaussian distribution
!       is used in IRANBIN the test may fail if P is close to 1
!       as the Gaussian distribution may not be a good approximation in
!       that case. RMAX in IRANBIN may be increased to avoid use of
!       Gaussian approximation
      PROB=1-GAMMAP(NU/2Q0,CHI/2)
      WRITE(6,54) PROB
      END

!       ------------------------------------------------------------

!	To calculate Gamma function for any real value of XG
!	Use GAMMAL for calculating the logarithm of Gamma function
!	which may be useful for large arguments or when argument is
!	close to a negative integer.
!
!	Required routines : None
 
      FUNCTION GAMMA(XG)
      IMPLICIT REAL*16(A-H,O-Z)
!	PIS is SQRT(2*PI)
      PARAMETER(PI=3.14159265358979323846Q0,PIS=2.5066282746310005024Q0)
      DIMENSION A(2),B(3),A1(6),B1(7)

!	The coefficients for rational function approximations
      DATA A/1.767971449569122937Q+00,  2.909421117928672645Q-01/
      DATA B/8.333333333333231537Q-02,  1.445531763554246280Q-01,
     1       2.012779361583001035Q-02/
      DATA A1/3.905731686764559737Q+03,  2.204952264401381785Q+03,
     1       -1.932467485468849660Q+03,  4.643360871045442213Q+02,
     1       -4.818088806916028754Q+01,  1.896853765546068169Q+00/
      DATA B1/3.918055655523400310Q+03, -1.088116266563809683Q+02,
     1        8.203258626193993149Q+02, -9.289402000761705906Q+01,
     1        6.521113026294866877Q+01, -6.090618615608719044Q+00,
     1        1.475909104740280784Q+00/
 
      X=ABS(XG)
      IF(X.GT.1000.0) THEN
!	Use asymptotic approximation (Stirling formula)
        GX=(1+1.Q0/(12*X)+1./(288*X*X)
     1         -139/(51840*X**3)-571./(2488320Q0*X**4))
        GAMMA=X**(X-0.5)*EXP(-X)*PIS*GX

      ELSE IF(X.GT.8.0) THEN
!	Use rational function approximation for Log(Gamma) 
        Y=1./X**2
        RMK=((B(3)*Y+B(2))*Y+B(1))/((A(2)*Y+A(1))*Y+1)
        GAMMA=X**(X-0.5)*EXP(-X)*PIS*EXP(RMK/X)

      ELSE IF(X.GE.2.0) THEN
!	Use rational function approximation for (Gamma) over [2,3]
!	after translating the range if necessary
        F1=1.0
        X1=X
2500    IF(X1.LE.3) GO TO 3000
        F1=F1*(X1-1)
        X1=X1-1
        GO TO 2500
3000    IF(X1.EQ.3) THEN
          GAMMA=F1*2
        ELSE IF(X1.EQ.2) THEN
          GAMMA=F1
        ENDIF

        FN=(((((B1(7)*X1+B1(6))*X1+B1(5))*X1+B1(4))*X1+B1(3))*X1+
     1          B1(2))*X1+B1(1)
        FD=(((((A1(6)*X1+A1(5))*X1+A1(4))*X1+A1(3))*X1+A1(2))*X1+
     1          A1(1))*X1+1
        GAMMA=F1*FN/FD
      ELSE IF(X.GT.0.0) THEN
!	Use rational function approximation for (Gamma) over [2,3]
!	after translating the range if necessary
        F1=1./X
        X1=X+1
        IF(X.LT.1) THEN
          F1=F1/X1
          X1=X1+1
        ENDIF
        IF(X1.EQ.2) GAMMA=F1

        FN=(((((B1(7)*X1+B1(6))*X1+B1(5))*X1+B1(4))*X1+B1(3))*X1+
     1          B1(2))*X1+B1(1)
        FD=(((((A1(6)*X1+A1(5))*X1+A1(4))*X1+A1(3))*X1+A1(2))*X1+
     1          A1(1))*X1+1
        GAMMA=F1*FN/FD
 
      ENDIF

      IF(XG.GT.0.0) RETURN
      IX=X
      IF(X.GT.IX) THEN
        GAMMA=PI/(XG*SIN(PI*X)*GAMMA)
      ELSE
        GAMMA=(-1)**IX/0.0
      ENDIF
 
      END

!       ------------------------------------------------------------

!	To calculate the Logarithm of Gamma function for a real argument
!	For negative values it give ln(abs(Gamma(x)))
!
!	Required routines : None
 
      FUNCTION GAMMAL(X)
      IMPLICIT REAL*16(A-H,O-Z)
      DIMENSION A(4),B(4),A1(6),B1(6)
!	PI2L=LOG(2*PI)/2
      PARAMETER (PI2L=0.918938533204672741Q0,PI=3.141592653589793238Q0)

!	The coefficients of rational function approximations
      DATA A/ 1.013142782275024216Q-2,  7.645657825398191944Q-1,
     1        3.381172379819461227Q-4,  1.595363637547538209Q-2/
      DATA B/0.0, 8.333333333338911768Q-2, 8.442856404442060242Q-4,
     1            6.093603832366013704Q-2/

      DATA A1/ 4.681163846241230144Q0,  3.208225429683256526Q0,
     1         5.145525793448859216Q-1, 1.581117883959157936Q-2,
     2        -6.398416804905407512Q-5, 5.264566254181773919Q-7/
      DATA B1/ 2.938038561191284576Q0,  1.489364948862436743Q0,
     1        -5.466291543917642961Q0,  1.972497734170110410Q-1,
     2         7.830146473241555157Q-1, 5.756753067834747499Q-2/
 
      T=ABS(X)
      IT=T
      IF(T.GE.10.0) THEN
!	Use asymptotic approximation for T>10
        Y=1/T
        FN=((B(4)*Y+B(3))*Y+B(2))*Y+B(1)
        FD=(((A(4)*Y+A(3))*Y+A(2))*Y+A(1))*Y+1
        GAMMAL=FN/FD-T+PI2L+(T-0.5)*LOG(T)
      ELSE
 
!	Use approximation for [4,5]
        FAC=0.0
        T1=T
        IF(T.LT.4) THEN
          DO I=IT,3
            FAC=FAC-LOG(T1)
            T1=T1+1
          ENDDO
        ELSE IF(T.GT.5) THEN
          DO I=IT,5,-1
            T1=T1-1
            FAC=FAC+LOG(T1)
          ENDDO
        ENDIF
        FN=((((B1(6)*T1+B1(5))*T1+B1(4))*T1+B1(3))*T1+B1(2))*T1+B1(1)
        FD=(((((A1(6)*T1+A1(5))*T1+A1(4))*T1+A1(3))*T1+A1(2))*T1+
     1          A1(1))*T1+1
        GAMMAL=FN/FD+FAC
      ENDIF

      IF(X.LE.0) THEN
        IF(T.GT.IT) THEN
          GAMMAL=LOG(PI)-LOG(ABS(SIN(PI*T)*T))-GAMMAL
        ELSE
          GAMMAL=(-1)**IT/0.0
        ENDIF
      ENDIF
 
      END

!       ------------------------------------------------------------

!	To calculate the incomplete Gamma function P(a,x) for
!	positive real arguments
!	It returns a value of -1 for a.le.0 or x<0
!
!	A : (input) Argument for the complete Gamma function
!	X : (input) Upper limit of integration defining the incomplete
!                       Gamma function
!
!	Required routines : GAMMA, GAMMAL

      FUNCTION GAMMAP(A,X)
      IMPLICIT REAL*16(A-H,O-Z)
      DIMENSION C(500),B(500)

      IF(A.LE.0.OR.X.LT.0) THEN
        GAMMAP=-1
        RETURN
      ENDIF

      IF(X.LT.3) THEN
!	Use power series
        S1=1
        T=A
        DO I=1,500
          T=-T*X/I
          S1=S1+T/(A+I)
          IF(ABS(T/(A+I)).LT.1.Q-14) EXIT
        ENDDO
        IF(A.LT.140) THEN
          GAMMAP=X**A*S1/GAMMA(A+1)
        ELSE
          GAMMAP=0.0
        ENDIF
      ELSE IF(A.LT.1.2Q0*X) THEN
!	Use continued fraction
        C(1)=1
        B(1)=X
        C(2)=1
        B(2)=X+1-A
        C1=C(2)/B(2)
        DO I=2,498,2
          C(I+1)=X*C(I)+(I/2)*C(I-1)
          B(I+1)=X*B(I)+(I/2)*B(I-1)
          C(I+2)=C(I+1)+(I/2+1-A)*C(I)
          B(I+2)=B(I+1)+(I/2+1-A)*B(I)
          IF(ABS(B(I+2)).GT.1.Q200) THEN
            C(I+1)=C(I+1)/1.Q200
            B(I+1)=B(I+1)/1.Q200
            C(I+2)=C(I+2)/1.Q200
            B(I+2)=B(I+2)/1.Q200
          ENDIF
          IF(ABS(B(I+2)).LT.1.Q-200) THEN
            C(I+1)=C(I+1)*1.Q200
            B(I+1)=B(I+1)*1.Q200
            C(I+2)=C(I+2)*1.Q200
            B(I+2)=B(I+2)*1.Q200
          ENDIF
          C2=C(I+2)/B(I+2)
          IF(ABS(C2-C1).LT.1.Q-12) EXIT
          C1=C2
        ENDDO
        G1=-X+A*LOG(X)+LOG(C2)-GAMMAL(A)
        GAMMAP=1-EXP(G1)
      ELSE
!	Use the power series for a>x
        S2=1
        T=1
        DO I=1,500
          T=T*X/(A+I)
          S2=S2+T
          IF(ABS(T).LT.1.Q-14) EXIT
        ENDDO
        G1=-X+A*LOG(X)+LOG(S2)-GAMMAL(A+1)
        GAMMAP=EXP(G1)
      ENDIF
      END

!       ------------------------------------------------------------

!       To generate random numbers with Binomial distribution 
!       For large values of mean it approximates the distribution by a
!       normal distribution with the same mean and standard deviation.
!
!       SEED : (input/output) real seed, it should be negative during
!              the first call and should not be modified between two
!              calls with the same N, P. If a new value of N or P is used
!              it should be reset to a negative value.
!       N   :  (input) Number of trials in the binomial distribution
!       P   :  (input) probability of the event
!       C   :  (output) Real array of length N used to store the
!              cumulative probability table for use in subsequent
!              calculations. This array should not be modified by the
!              user. This is used only if the mean N*P<RMAX and there
!              is no underflow while calculating (1-P)**N
!
!       Required routines : None
!
!       If P is close to 1 the approximation by Gaussian may not be very
!       good and in that case increasing RMAX may help although it will
!       take longer time to generate the numbers.

      FUNCTION IRANBIN(SEED,N,P,C)
      IMPLICIT REAL*16(A-H,O-Z)
      DIMENSION C(*)
      PARAMETER(AM=2147483648Q0,A=45875,AC=453816693Q0,AN=2147483647Q0)
      PARAMETER(PI=3.14159265358979323846264338327950288Q0,RMAX=100)

      RMU=N*P
      SIG=SQRT(RMU*(1-P))
      P0=(1-P)**N

!       For large mu of if P0 gives underflow use normal distribution
      IF(RMU.GT.RMAX.OR.P0.EQ.0.0) THEN
!       Initialise the seed during first call
        IF(SEED.LT.0.0) SEED=MOD(ABS(SEED),AM)
        R1=MOD(A*SEED+AC,AM)
        IF(SEED.EQ.0.0) SEED=0.1Q0
        RANGAU=SQRT(2.Q0*LOG(AN/SEED))*COS(2.0*PI*R1/AN)
        SEED=R1
        IRANBIN=RMU+SIG*RANGAU+0.5
        IF(IRANBIN.LT.0) IRANBIN=0
        IF(IRANBIN.GT.N) IRANBIN=N
      ELSE
!       Initialise the array during the first call
        IF(SEED.LT.0) THEN
          C(1)=P0
          DO I=2,N+1
            P0=P0*P*(N-I+2)/((1-P)*(I-1))
            C(I)=C(I-1)+P0
          ENDDO
          SEED=ABS(SEED)
        ENDIF

!       generate the random number
        SEED=MOD(SEED*A+AC,AM)
        R=SEED/AN
        DO I=1,N+1
          IF(R.LT.C(I)) EXIT
        ENDDO
        IF(I.GT.N+1) I=N+1
        IRANBIN=I-1
      ENDIF
      END
